File: herencia-dto/SergioDeIscar-herencia-java/SergioDeIscar-herencia-java/src/main/java/org/example/services/storage/persona/PersonaFileXml.java
package org.example.services.storage.persona;

import org.example.config.AppConfig;
import org.example.dto.PersonaDto;
import org.example.dto.PersonasDto;
import org.example.models.Persona;
import org.simpleframework.xml.core.Persister;

import java.io.File;
import java.util.List;

public class PersonaFileXml implements PersonaStorageService{
    //Singleton
    private static PersonaFileXml instance = null;
    public static PersonaFileXml getInstance() {
        if (instance == null) {
            instance = new PersonaFileXml();
        }
        return instance;
    }

    private final String localFile = AppConfig.getInstance().getAppData() + File.separator + "personas.xml";

    @Override
    public List<Persona> saveAll(List<Persona> elements) {
        File file = new File(localFile);
        if (file.exists() && !file.canWrite() ) return null;
        Persister serializer = new Persister();
        try {
            serializer.write(new PersonasDto(elements.stream().map(Persona::toDto).toList()), file);
            return elements;
        } catch (Exception e) {
            System.out.println("Error al escribir en el fichero");
        }
        return null;
    }

    @Override
    public List<Persona> loadAll() {
        File file = new File(localFile);
        if(!file.exists() || !file.canRead()) return null;
        try {
            PersonasDto personasDto = new Persister().read(PersonasDto.class, file);
            return personasDto.getPersonas().stream().map(PersonaDto::toClass).toList();
        } catch (Exception e) {
            System.out.println("Error al leer en el fichero");
        }
        return null;
    }
}

File: herencia-dto/SergioDeIscar-herencia-kotlin/SergioDeIscar-herencia-kotlin/src/main/kotlin/service/storage/persona/PersonaFileXml.kt
package service.storage.persona

import config.AppConfig
import dto.PersonasDto
import mappers.toClass
import mappers.toDto
import models.Persona
import mu.KotlinLogging
import org.simpleframework.xml.core.Persister
import java.io.File

private val logger = KotlinLogging.logger {}

object PersonaFileXml: PersonaStorageService{
    private val localFile = "${AppConfig.APP_DATA}${File.separator}personas.xml"

    override fun saveAll(elements: List<Persona>): List<Persona> {
        logger.debug { "PersonaFileXml ->\tsaveAll" }
        val file = File(localFile)
        if (file.exists() && !file.canWrite() ) return emptyList()
        val serializer = Persister()
        serializer.write(PersonasDto(elements.map { it.toDto() }), file)
        return elements
    }

    override fun loadAll(): List<Persona> {
        logger.debug { "PersonaFileXml ->\tloadAll" }
        val file = File(localFile)
        if(!file.exists() || !file.canRead()) return emptyList()
        return Persister().read(PersonasDto::class.java, file).personas.map { it.toClass() }
    }
}